virtual-display-driver-cli/src/client.rs

```diff
@@ -0,0 +1,113 @@
+use std::io::Write as _;
+
+use driver_ipc::Monitor;
+use eyre::Context;
+use win_pipes::{NamedPipeClientReader, NamedPipeClientWriter};
+
+pub struct Client {
+    reader: NamedPipeClientReader,
+    writer: NamedPipeClientWriter,
+}
+
+impl Client {
+    pub fn connect() -> eyre::Result<Self> {
+        let (reader, writer) = win_pipes::NamedPipeClientOptions::new("virtualdisplaydriver")
+            .wait()
+            .access_duplex()
+            .mode_message()
+            .create()?;
+
+        Ok(Self { reader, writer })
+    }
+
+    pub fn list(&mut self) -> eyre::Result<Vec<Monitor>> {
+        self.send_command(&driver_ipc::Command::RequestState)?;
+        let response = self.receive_command()?;
+
+        match response {
+            driver_ipc::Command::ReplyState(state) => Ok(state),
+            _ => eyre::bail!("received unexpected reply from driver pipe"),
+        }
+    }
+
+    pub fn get(&mut self, id: driver_ipc::Id) -> eyre::Result<Monitor> {
+        let monitors = self.list()?;
+        let monitor = monitors.into_iter().find(|monitor| monitor.id == id);
+
+        match monitor {
+            Some(monitor) => Ok(monitor),
+            None => eyre::bail!("no virtual monitor with ID {} found", id),
+        }
+    }
+
+    pub fn notify(&mut self, monitors: Vec<driver_ipc::Monitor>) -> eyre::Result<()> {
+        let command = driver_ipc::Command::DriverNotify(monitors);
+
+        self.send_command(&command)?;
+
+        Ok(())
+    }
+
+    pub fn remove(&mut self, ids: Vec<driver_ipc::Id>) -> eyre::Result<()> {
+        let command = driver_ipc::Command::DriverRemove(ids);
+
+        self.send_command(&command)?;
+
+        Ok(())
+    }
+
+    pub fn remove_all(&mut self) -> eyre::Result<()> {
+        let command = driver_ipc::Command::DriverRemoveAll;
+
+        self.send_command(&command)?;
+
+        Ok(())
+    }
+
+    pub fn new_id(&mut self, preferred_id: Option<driver_ipc::Id>) -> eyre::Result<driver_ipc::Id> {
+        let monitors = self.list()?;
+
+        match preferred_id {
+            Some(id) => {
+                if monitors.iter().any(|monitor| monitor.id == id) {
+                    eyre::bail!("monitor with ID {} already exists", id);
+                }
+
+                Ok(id)
+            }
+            None => {
+                let max_id = monitors.iter().map(|monitor| monitor.id).max();
+                let id = match max_id {
+                    Some(id) => id + 1,
+                    None => 0,
+                };
+                Ok(id)
```

Comment:
I definitely went with this implementation because it was fast and easy, not because it was the most intuitive 😅

The other option would of course be to use the smallest available ID. I would kind of hope most users wouldn't care too much about how IDs get set, but the smallest available would probably be clearer to be honest, so I think it would make sense to switch to that.

Reply:
_Marking as resolved, this will be handled in a follow-up_